src/Microsoft.VisualStudio.Threading/JoinableTaskDependencyGraph.cs

```diff
@@ -724,61 +873,27 @@ private static void RemoveDependingSynchronousTaskFrom(IReadOnlyList<IJoinableTa
                     RemoveDependingSynchronousTask(task, syncTask, reachableNodes, ref remainNodes);
                 }
 
-                if (!force && remainNodes is object && remainNodes.Count > 0)
+                if (remainNodes is object && remainNodes.Count > 0)
                 {
-                    // a set of tasks may form a dependent loop, so it will make the reference count system
-                    // not to work correctly when we try to remove the synchronous task.
-                    // To get rid of those loops, if a task still tracks the synchronous task after reducing
-                    // the reference count, we will calculate the entire reachable tree from the root.  That will
-                    // tell us the exactly tasks which need track the synchronous task, and we will clean up the rest.
-                    reachableNodes = new HashSet<IJoinableTaskDependent>();
-                    var syncTaskItem = (IJoinableTaskDependent)syncTask;
-                    ComputeSelfAndDescendentOrJoinedJobsAndRemainTasks(syncTaskItem, reachableNodes, remainNodes);
+                    if (force)
+                    {
+                        Assumes.True(reachableNodes!.Count == 0);
```

Comment:
`reachableNodes!`
I see why you can predict that this will never be null. Consider changing the `if` check above it from `force` to `reachableNodes is object` as it appears to be an equivalent condition and would then allow you to remove the null forgiveness operator.

Reply:
I added an Assumes.NotNull. I agree with you that changing the condition would make this unnecessary. But it also makes the condition indirect, and hard to make sense of why we can remove all unreachable nodes here.